server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -141,6 +166,30 @@ public void afterRefresh(boolean didRefresh) {
         }
     }
 
+    private boolean isRefreshAfterCommit() throws IOException {
+        String lastCommittedLocalSegmentFileName = SegmentInfos.getLastCommitSegmentsFileName(storeDirectory);
+        return (lastCommittedLocalSegmentFileName != null
+            && !remoteDirectory.containsFile(lastCommittedLocalSegmentFileName, getChecksumOfLocalFile(lastCommittedLocalSegmentFileName)));
+    }
+
+    String uploadSegmentInfosSnapshot(String latestSegmentsNFilename, SegmentInfos segmentInfosSnapshot) throws IOException {
+        long processedLocalCheckpoint = indexShard.getEngine().getProcessedLocalCheckpoint();
```

Comment:
I think the logic for segment upload is growing in the Listener, would suggest moving all the upload logic to a dedicated class.

Reply:
Got it. Looking at the remote translog PR, I think we can use the similar concept of TransferFileSnapshot that is used [here](https://github.com/opensearch-project/OpenSearch/pull/5392). 

We are already tracking a similar ask here: https://github.com/opensearch-project/OpenSearch/issues/4198

Reply:
How do we ensure that the processed local checkpoint corresponds to the latest segment info as in can they go out of sync?

Reply:
We don't have to keep them in sync as long as processed local checkpoint is <= max sequence number that is part of segment files of SegmentInfos.

I think you are referring to fetching `processedCheckpoint` after fetching SegmentInfos. This is a bug. I will fix it. We want checkpoint to be fetched before.

Reply:
@Bukhtawar Fixed the bug in the latest commit. Please review.